scanner/base.go

```diff
@@ -794,16 +795,39 @@ func (d *DummyFileInfo) IsDir() bool { return false }
 //Sys is
 func (d *DummyFileInfo) Sys() interface{} { return nil }
 
+func (l *base) buildWpCliCmd(wpCliArgs string, suppressStderr bool) string {
+	cmd := ""
+	if !l.ServerInfo.WordPress.NoSudo {
+		cmd = fmt.Sprintf("sudo -u %s -i -- ", l.ServerInfo.WordPress.OSUser)
+	}
+	cmd += fmt.Sprintf("%s %s --path=%s --allow-root",
```

Comment:
If `NoSudo = true`, then the User who executes the command will always be the Scan User, shouldn't it run on WordPress.OSUser?

Reply:
sudoが実行できない環境(さくらレンタルサーバ)向けの修正なのでScan User(ServerInfo.User=ssh接続ユーザ)でwp-cliコマンドを実行してます。

Reply:
(a) ServerInfo.User != ServerInfo.WordPress.OSUser
この場合はsudo不要というのが最初の実装の意図です。
(b) !ServerInfo.WordPress.NoSudo 
(a)と比べ、既設定でServerInfo.User = ServerInfo.WordPress.OSUser があってもsudoを省略しない／同じコマンドが発行されるのが利点。 
(c) (b)は止めて ServerInfo.WordPress.OSUser が未定義の場合にsudoしないように変更する？

Reply:
I understand that the rental server provided by Sakura Internet has such restrictions.
However, this part applies to servers other than those provided by Sakura Internet, so it should take into account the general user environment.

For this problem, we should consider the case where neither `ServerInfo.User` nor `ServerInfo.WordPress.OSUser` has Sudo privileges.
In this case, the user who should execute the command should be `ServerInfo.WordPress.OSUser`.
If `ServerInfo.User` equals `ServerInfo.WordPress.OSUser`, the command is executed as is.
If different, the command should be run as `ServerInfo.WordPress.OSUser`.

Reply:
上記のケースで wp-cliコマンド実行ユーザを `ServerInfo.WordPress.OSUser` にする場合の実装アイデアはありますか。
ssh接続ユーザを `ServerInfo.WordPress.OSUser` にするのを考えたのですが破壊的変更になりそうです。

Reply:
> ssh接続ユーザを ServerInfo.WordPress.OSUser にするのを考えたのですが破壊的変更になりそうです。

Do you mean to force `ServerInfo.User` to change to `ServerInfo.WordPress.OSUser`?
If that is what you think, then it is unacceptable.

Reply:
そうですよね。何かアイデアがあればお願いします。

Reply:
I don't currently have a WordPress environment, so I haven't been able to verify this patch, but here's what it looks like.
```diff
:100644 100644 861e375 0000000 M	scanner/base.go

diff --git a/scanner/base.go b/scanner/base.go
index 861e375..2bc5864 100644
--- a/scanner/base.go
+++ b/scanner/base.go
@@ -795,15 +795,15 @@ func (d *DummyFileInfo) IsDir() bool { return false }
 func (d *DummyFileInfo) Sys() interface{} { return nil }
 
 func (l *base) buildWpCliCmd(wpCliArgs string, suppressStderr bool, shell string) string {
-	cmd := ""
+	cmd := fmt.Sprintf("%s %s --path=%s", l.ServerInfo.WordPress.CmdPath, wpCliArgs, l.ServerInfo.WordPress.DocRoot)
 	if !l.ServerInfo.WordPress.NoSudo {
-		cmd = fmt.Sprintf("sudo -u %s -i -- ", l.ServerInfo.WordPress.OSUser)
+		cmd = fmt.Sprintf("sudo -u %s -i -- %s --allow-root", l.ServerInfo.WordPress.OSUser, cmd)
+	} else {
+		if l.ServerInfo.User != l.ServerInfo.WordPress.OSUser {
+			cmd = fmt.Sprintf("su %s -c '%s'", l.ServerInfo.WordPress.OSUser, cmd)
+		}
 	}
-	cmd = fmt.Sprintf("%s %s %s --path=%s --allow-root",
-		cmd,
-		l.ServerInfo.WordPress.CmdPath,
-		wpCliArgs,
-		l.ServerInfo.WordPress.DocRoot)
+
 	if suppressStderr {
 		switch shell {
 		case "csh", "tcsh":
@@ -819,16 +819,15 @@ func (l *base) scanWordPress() error {
 	if l.ServerInfo.WordPress.IsZero() || l.ServerInfo.Type == constant.ServerTypePseudo {
 		return nil
 	}
-	// determine shell
-	r := l.exec("echo $SHELL", noSudo)
-	if !r.isSuccess() {
-		return xerrors.Errorf("Failed to exec echo $SHELL: %s", r)
+
+	shell, err := l.detectShell()
+	if err != nil {
+		return xerrors.Errorf("Failed to detect shell. err: %w", err)
 	}
-	shell := filepath.Base(strings.TrimSpace(r.Stdout))
 
 	l.log.Info("Scanning WordPress...")
 	cmd := l.buildWpCliCmd("core version", false, shell)
-	if r = exec(l.ServerInfo, cmd, noSudo); !r.isSuccess() {
+	if r := exec(l.ServerInfo, cmd, noSudo); !r.isSuccess() {
 		return xerrors.Errorf("Failed to exec `%s`. Check the OS user, command path of wp-cli, DocRoot and permission: %#v", cmd, l.ServerInfo.WordPress)
 	}
 
@@ -840,6 +839,32 @@ func (l *base) scanWordPress() error {
 	return nil
 }
 
+func (l *base) detectShell() (string, error) {
+	if r := l.exec("printenv SHELL", noSudo); r.isSuccess() {
+		if t := strings.TrimSpace(r.Stdout); t != "" {
+			return filepath.Base(t), nil
+		}
+	}
+
+	if r := l.exec(fmt.Sprintf(`grep "^%s" /etc/passwd | awk -F: '/%s/ { print $7 }'`, l.ServerInfo.User, l.ServerInfo.User), noSudo); r.isSuccess() {
+		if t := strings.TrimSpace(r.Stdout); t != "" {
+			return filepath.Base(t), nil
+		}
+	}
+
+	if isLocalExec(l.ServerInfo.Port, l.ServerInfo.Host) {
+		if r := l.exec("ps -p $$ | tail +2 | awk '{print $NF}'", noSudo); r.isSuccess() {
+			return strings.TrimSpace(r.Stdout), nil
+		}
+
+		if r := l.exec("ps -p %self | tail +2 | awk '{print $NF}'", noSudo); r.isSuccess() {
+			return strings.TrimSpace(r.Stdout), nil
+		}
+	}
+
+	return "", xerrors.New("shell cannot be determined")
+}
+
 func (l *base) detectWordPress(shell string) (*models.WordPressPackages, error) {
 	ver, err := l.detectWpCore(shell)
 	if err != nil {

```

Reply:
>Note that if ServerInfo.User and ServerInfo.WordPress.OSUser are different, a password is required.
It may be a good idea to verify in advance whether it is possible to switch users without passwords.

パスワードなしにする場合rootユーザでの実行かpam.dでの許可が必要になると思います。
その設定ができる環境は、現在ドキュメントにあるsudoで対応できそうにみえます。

Reply:
上記コードマージしてbash,csh環境で動作確認しました。

Reply:
> > Note that if ServerInfo.User and ServerInfo.WordPress.OSUser are different, a password is required.
> > It may be a good idea to verify in advance whether it is possible to switch users without passwords.
> 
> パスワードなしにする場合rootユーザでの実行かpam.dでの許可が必要になると思います。 その設定ができる環境は、現在ドキュメントにあるsudoで対応できそうにみえます。

To set a no password, it is necessary to execute the command with sudo privilege or as root user. So, realistically, if `ServerInfo.User` and `ServerInfo.WordPress.OSUser` are different, and if the environment is such that the user can be switched without password, do you think it is sufficient to run it as `NoSudo = false` to begin with?
By the way, what do you mean by documentation(`ドキュメント`)?

Reply:
A generic solution should be considered when `ServerInfo.User` and `ServerInfo.WordPress.OSUser` are different and `ServerInfo.User` does not have Sudo privileges.

I do not want to write Password in Config.toml, so I think it should be an error if a user cannot be switched without a password.
For this reason, we should check whether a password is required or not.


Are you going to implement this?

Reply:
実装すすめてません。

sudo(sudoers)の設定の際に以下のドキュメント参考にしました。
https://help.vuls.biz/manual/scan/wordpress/#%E3%82%B9%E3%82%AD%E3%83%A3%E3%83%B3%E5%AF%BE%E8%B1%A1%E3%82%B5%E3%83%BC%E3%83%90%E4%B8%8A%E3%81%AEsudoers%E3%81%AE%E8%A8%AD%E5%AE%9A

sudo実行権限と、ServerInfo.UserのServerInfo.WordPress.OSUser一致／不一致の組み合わせは4パターン
(1),(2)が現状できる設定です。

(1) `ServerInfo.User` sudo 実行権限あり, `ServerInfo.User` = `ServerInfo.WordPress.OSUser` 
(2) `ServerInfo.User` sudo 実行権限あり, `ServerInfo.User` != `ServerInfo.WordPress.OSUser` 
(3) `ServerInfo.User` sudo 実行権限なし, `ServerInfo.User` = `ServerInfo.WordPress.OSUser` 
(4) `ServerInfo.User` sudo 実行権限なし, `ServerInfo.User` != `ServerInfo.WordPress.OSUser`

>A generic solution should be considered when ServerInfo.User and ServerInfo.WordPress.OSUser are different and ServerInfo.User does not have Sudo privileges.

(4)が必要なケースが思い当たらないので、`ServerInfo.WordPress.NoSudo` かつ `ServerInfo.User` `ServerInfo.WordPress.OSUser` が不一致の場合は、設定のエラー（設定できない組み合わせ）とするのはどうでしょうか。

Reply:
For example, there may be a case where the execution of the sudo command is being audited in the log, so that the command should be executed by an appropriate user with minimal privileges, without using sudo as much as possible.

Reply:
I will implement this case.
I will also write this document(https://github.com/future-architect/vuls/pull/1523#issuecomment-1232464968).
Thank you for the PR.

Reply:
ありがとうございます。
理解、実装ともに力不足の状態でPRをだしてしまい、お手数をかけてしまって申し訳有りません。

Reply:
Without your PR suggestion, I would never have thought about such a case.
Additionally, this PR will be a good cover for the content of PRs that have been stopped in WIP status.

Thank you for your continued support of Vuls.

Reply:
ありがとうございます。vuls使用させて頂きとても役立っております。
後知恵ですが、NoSudoという名前よくなかったです。ansibleの [become_method](https://docs.ansible.com/ansible/latest/user_guide/become.html) のように権限昇格ツールの名前入れる名前にして "none" "sudo" "su" のようにしておけばよかったと思いました。（デフォルト"sudo"で互換性を考慮する）